### PR TITLE
Add caution for using memory-backed emptydir

### PR DESCRIPTION
**Pull request by shu-mutou**
To use default size of memory-backed emptydir, some risks exist. So this adds cautions for the risks.

Ref: https://github.com/kubernetes/kubernetes/issues/119611

### CODE REVIEW

**Comment by T-Lakshmi** [comment]
[Preview page](https://deploy-preview-44949--kubernetes-io-main-staging.netlify.app/docs/concepts/storage/volumes/)

LGTM for docs

**Comment by shu-mutou** [comment]
[Preview page](https://deploy-preview-44949--kubernetes-io-main-staging.netlify.app/docs/concepts/storage/volumes/#emptydir)

File: content/en/docs/concepts/storage/volumes.md [hunk 1]
```diff
@@ -255,11 +255,40 @@ If that is filled up from another source (for example, log files or image
 overlays), the `emptyDir` may run out of capacity before this limit.
 
 {{< note >}}
-If the `SizeMemoryBackedVolumes` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) is enabled,
-you can specify a size for memory backed volumes.  If no size is specified, memory
+The `SizeMemoryBackedVolumes` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) is enabled by default since v1.22,
+so you can specify a size for memory backed volumes.  If no size is specified, memory
 backed volumes are sized to node allocatable memory.
```
Comment (sftim) [inline content/en/docs/concepts/storage/volumes.md:260]: Try this:
```suggestion
You can specify a size for memory backed volumes, provided that the
`SizeMemoryBackedVolumes` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
enabled in your cluster (this has been beta, and active by default, since the Kubernetes 1.22 release).
If you don't specify a volume size, memory
```

File: content/en/docs/concepts/storage/volumes.md [hunk 2]
```diff
@@ -255,11 +255,40 @@ If that is filled up from another source (for example, log files or image
...
+{{< caution >}}
+If the user does not specify a `sizeLimit` for `emptyDir`, user can create a
+memory backed `emptyDir` with node allocatable size unintentionally.  Also,
+user can create multiple memory backed `emptyDir`s on the same node, each set
+to a node allocatable size.  And what Kubernetes sees as memory usage is the
+amount that is actually consuming the memory　backed `emptyDir`, so even after
+you create a memory-backed `emptyDir` with node allocatable size, the other Pods
+can be scheduled to the same node.  As a result, multiple memory backed
+`emptyDir`s with node allocatable size can be created on the same node.
+
```
Comment (sftim) [inline content/en/docs/concepts/storage/volumes.md:272]: Consider ending the `caution` block here, with some advise to continue reading the page to learn  about managing those risks.

Also, how about moving the `caution` earlier in the page (eg: around line 224)?

File: content/en/docs/concepts/storage/volumes.md [hunk 3]
```diff
@@ -1085,7 +1114,7 @@ For more information on how to develop a CSI driver, refer to the
 
 CSI node plugins need to perform various privileged
 operations like scanning of disk devices and mounting of file systems. These operations
-differ for each host operating system. For Linux worker nodes, containerized CSI node
```
Comment (sftim) [inline content/en/docs/concepts/storage/volumes.md:1088]: Please either:
- omit the whitespace change
- move it to its own commit

File: content/en/docs/concepts/storage/volumes.md [hunk 4]
```diff
@@ -1107,7 +1136,7 @@ configuration changes to existing Storage Classes, PersistentVolumes or Persiste
 Existing PVs created by a in-tree volume plugin can still be used in the future without any configuration
 changes, even after the migration to CSI is completed for that volume type, and even after you upgrade to a
 version of Kubernetes that doesn't have compiled-in support for that kind of storage.
-
```
Comment (sftim) [inline content/en/docs/concepts/storage/volumes.md:1110]: Please either:
- omit the whitespace change
- move it to its own commit

File: content/en/docs/concepts/storage/volumes.md [hunk 5]
```diff
@@ -201,7 +201,7 @@ keyed with `log_level`.
 
 * A container using a ConfigMap as a [`subPath`](#using-subpath) volume mount will not
   receive ConfigMap updates.
-
```
Comment (sftim) [inline content/en/docs/concepts/storage/volumes.md:204]: Please either:
- omit the whitespace change
- move it to its own commit

File: content/en/docs/concepts/storage/volumes.md [hunk 6]
```diff
@@ -12,12 +12,12 @@ weight: 10
...
-files that were created or modified during the lifetime of the container are lost. 
-During a crash, kubelet restarts the container with a clean state. 
-Another problem occurs when multiple containers are running in a `Pod` and 
-need to share files. It can be challenging to setup 
+non-trivial applications when running in containers. One problem occurs when
+a container crashes or is stopped. Container state is not saved so all of the
+files that were created or modified during the lifetime of the container are lost.
+During a crash, kubelet restarts the container with a clean state.
+Another problem occurs when multiple containers are running in a `Pod` and
+need to share files. It can be challenging to setup
```
Comment (sftim) [inline content/en/docs/concepts/storage/volumes.md:20]: We'd prefer whitespace fixes to land in their own commit.

**Review by sftim — COMMENTED** [review]
Thanks

It's a big caution! How about making a smaller more prominent callout, that advises readers to read up about how and why to enforce size limits?

**Comment by shu-mutou** [comment]
@sftim Thank you for your reviewing!
I'll update this PR in this week.

File: content/en/docs/concepts/storage/volumes.md (hunk 2 above)
Comment (shu-mutou) [reply content/en/docs/concepts/storage/volumes.md:272]: I think an explanation about the whole emptyDir is needed before talking about memory backed volume. Instead of moving `caution` block earlier, I made a subsection about memory backed emptyDir, what do you think?

**Comment by shu-mutou** [comment]
[Preview page](https://deploy-preview-44949--kubernetes-io-main-staging.netlify.app/docs/concepts/storage/volumes/#before-using-memory-backed-emptydir)

File: content/en/docs/concepts/storage/volumes.md [hunk 7]
```diff
@@ -254,12 +254,52 @@ storage](/docs/concepts/configuration/manage-resources-containers/#setting-reque
 If that is filled up from another source (for example, log files or image
 overlays), the `emptyDir` may run out of capacity before this limit.
 
+#### Before using memory backed `emptyDir`
```
Comment (sftim) [inline content/en/docs/concepts/storage/volumes.md:257]: I (fairly strongly) recommend:
```suggestion
#### Considerations for memory backed `emptyDir` volumes {#emptydir-memory-backed}
```

**Review by sftim — COMMENTED** [review]
This would LGTM with one further change

**Comment by shu-mutou** [comment]
@sftim  Thank you for reviewing!

**Comment by sftim** [comment]
/lgtm

Thanks

**Comment by shu-mutou** [comment]
@dashpole Could you review this?

File: content/en/docs/concepts/storage/volumes.md [hunk 8]
```diff
@@ -254,12 +254,52 @@ storage](/docs/concepts/configuration/manage-resources-containers/#setting-reque
...
+you create a memory backed `emptyDir` with node allocatable size, the other Pods
+can be scheduled to the same node.  As a result, multiple memory backed
+`emptyDir`s with node allocatable size can be created on the same node.
+{{< /caution >}}
+
+If you use memory backed `emptyDir` without appropriate settings, the following
+risks may occur:
+
+* A single Pod with memory backed `emptyDir` may unintentionally occupy node's memory.
+* If multiple memory backed `emptyDir`s are created on the same mode, their total
```
Comment (dashpole) [inline content/en/docs/concepts/storage/volumes.md:281]: This isn't really a problem with memory-backed empty-dir volumes, right?  If you schedule pods with memory limits higher than memory requests, you can run into this scenario.

**Comment by sftim** [comment]
Thanks @dashpole 

The [review](https://github.com/kubernetes/website/pull/44949#pullrequestreview-1994449590) makes sense, and to emphasize that feedback:
/lgtm cancel

**Comment by shu-mutou** [comment]
@dashpole @sftim Thank you for your reviewings!!
I'll consider moving to the Resource Management documentation. It will take some time....

File: content/en/docs/concepts/storage/volumes.md [hunk 9]
```diff
@@ -254,12 +254,52 @@ storage](/docs/concepts/configuration/manage-resources-containers/#setting-reque
...
-backed volumes are sized to node allocatable memory.
+You can specify a size for memory backed volumes, provided that the `SizeMemoryBackedVolumes`
+[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
+enabled in your cluster (this has been beta, and active by default, since the Kubernetes 1.22 release).
+If you don't specify a volume size, memory backed volumes are sized to node allocatable memory.
 {{< /note>}}
 
+{{< caution >}}
+If you do not specify a `sizeLimit` for `emptyDir`, you can create a
+memory backed `emptyDir` with node allocatable size unintentionally.  Also,
```
[the remark this answers is not in the document]
Comment (shu-mutou) [reply content/en/docs/concepts/storage/volumes.md]: Added description about `nor a memory limit`. Also, added descriptions about purpose for memory limit and sizeLimit.

File: content/en/docs/concepts/storage/volumes.md [hunk 10]
```diff
@@ -254,12 +254,52 @@ storage](/docs/concepts/configuration/manage-resources-containers/#setting-reque
...
+amount that is actually consuming the memory　backed `emptyDir`, so even after
+you create a memory backed `emptyDir` with node allocatable size, the other Pods
+can be scheduled to the same node.  As a result, multiple memory backed
+`emptyDir`s with node allocatable size can be created on the same node.
+{{< /caution >}}
+
+If you use memory backed `emptyDir` without appropriate settings, the following
+risks may occur:
+
+* A single Pod with memory backed `emptyDir` may unintentionally occupy node's memory.
```
[the remark this answers is not in the document]
Comment (shu-mutou) [reply content/en/docs/concepts/storage/volumes.md]: Yes. However, using memory as a process work area and using it as a volume have different purposes, so I would like to describe the points to be especially careful about when using memory as a volume.

File: content/en/docs/concepts/storage/volumes.md (hunk 8 above)
Comment (shu-mutou) [reply content/en/docs/concepts/storage/volumes.md]: Yes, ditto. This is also the case when memory request and memory limit are not set, so these are added as the conditions mentioned above.

**Comment by shu-mutou** [comment]
@dashpole PTAL, thanks!!

**Comment by shu-mutou** [comment]
@dashpole @sftim Could you review this?

**Comment by shu-mutou** [comment]
@dashpole @sftim @thockin Could you please review this again?

File: content/en/docs/concepts/configuration/manage-resources-containers.md [hunk 11]
```diff
@@ -237,6 +238,73 @@ are available in your cluster, then Pod resource usage can be retrieved either
 from the [Metrics API](/docs/tasks/debug/debug-cluster/resource-metrics-pipeline/#metrics-api)
 directly or from your monitoring tools.
 
+### Considerations for memory backed `emptyDir` volumes {#memory-backed-emptydir}
+
+{{< caution >}}
+If you do not specify a `sizeLimit` for `emptyDir` nor a memory limits
```
Comment (thockin) [inline content/en/docs/concepts/configuration/manage-resources-containers.md:244]: nit "I'd say "limit" here rather than "limits"

File: content/en/docs/concepts/configuration/manage-resources-containers.md [hunk 12]
```diff
@@ -237,6 +238,73 @@ are available in your cluster, then Pod resource usage can be retrieved either
...
+(`spec.containers[].resources.limits.memory`) for pod, you can create a memory
```
Comment (thockin) [inline content/en/docs/concepts/configuration/manage-resources-containers.md:245]: ...your pod... ?

File: content/en/docs/concepts/configuration/manage-resources-containers.md [hunk 13]
```diff
@@ -237,6 +238,73 @@ are available in your cluster, then Pod resource usage can be retrieved either
...
+backed `emptyDir` with node allocatable size unintentionally. Also, user can
```
Comment (thockin) [inline content/en/docs/concepts/configuration/manage-resources-containers.md:246]: propose rephrasing:

```
If you do not specify a `sizeLimit` for an `emptyDir` volume, that volume may consume up to that pod's memory limit (`Pod.spec.containers[].resources.limits.memory`).  If you do not set a memory limit, the pod has no upper bound on memory consumption, and can consume all available memory on the node.  Kubernetes schedules pods based on resource requests (`Pod.spec.containers[].resources.requests`) and will not consider memory usage above the request when deciding if another pod can fit on a given node.  This can result in a denial of service and cause the OS to do out-of-memory (OOM) handling.
```

I don't think the aspect of multiple emptyDirs is significant - even one can DoS a machine.

File: content/en/docs/concepts/configuration/manage-resources-containers.md [hunk 14]
```diff
@@ -237,6 +238,73 @@ are available in your cluster, then Pod resource usage can be retrieved either
...
+is actually consuming the memory　backed `emptyDir`, so even after you create a
+memory backed `emptyDir` with node allocatable size, the other Pods can be
+scheduled to the same node.  As a result, multiple memory backed `emptyDir`s
+with node allocatable size can be created on the same node.
+{{< /caution >}}
+
+From the perspective of memory management, there are some similarities between
+when a process uses memory as a work area and when using memory-backed
+`emptyDir`. But when using memory as a volume like memory-backed `emptyDir`,
+there are additional points below that you should be careful.
```
Comment (thockin) [inline content/en/docs/concepts/configuration/manage-resources-containers.md:258]: ...careful of.

or

...consider.

File: content/en/docs/concepts/configuration/manage-resources-containers.md [hunk 15]
```diff
@@ -237,6 +238,73 @@ are available in your cluster, then Pod resource usage can be retrieved either
...
+
+* Files stored on a memory-backed volume are almost entirely managed by the
+  user application. Unlike when used as a work area for a process, you can not
+  rely on like garbage collection by the runtime for language.
```
Comment (thockin) [inline content/en/docs/concepts/configuration/manage-resources-containers.md:262]: drop "like" or change to "things like"

change "garbage collection by the runtime for language" to "language-level garbage collection" ?

File: content/en/docs/concepts/configuration/manage-resources-containers.md [hunk 16]
```diff
@@ -237,6 +238,73 @@ are available in your cluster, then Pod resource usage can be retrieved either
...
+* The purpose of writing files to a volume is to save data or pass it between
+  applications. Deletion of files may not be considered or may depend on other
```
Comment (thockin) [inline content/en/docs/concepts/configuration/manage-resources-containers.md:264]: The first sentence is good, the rest is a little twisty.  Suggest:

```
...between applications.  Neither Kubernetes nor the OS may automatically delete files from a volume, so memory used by those files can not be reclaimed when the system or the pod are under memory pressure.
```

File: content/en/docs/concepts/configuration/manage-resources-containers.md [hunk 17]
```diff
@@ -237,6 +238,73 @@ are available in your cluster, then Pod resource usage can be retrieved either
...
+  applications, then tend to persist for long time as files accumulate or are
+  not deleted as intended.
+* The memory-backed `emptyDir` is useful when writing large amounts of logs
```
Comment (thockin) [inline content/en/docs/concepts/configuration/manage-resources-containers.md:267]: simpler:

```
A memory-backed `emptyDir` is useful because of its performance, but memory is generally much smaller in size and much higher in cost than other storage media, such as disks or SSDs.  Using large amounts of memory for `emptyDir` volumes may affect  the normal operation of your pod or of the whole node, so should be used carefully.
```

File: content/en/docs/concepts/configuration/manage-resources-containers.md [hunk 18]
```diff
@@ -237,6 +238,73 @@ are available in your cluster, then Pod resource usage can be retrieved either
...
+  continuously or passing large image files because of its performance.
+  However, memory is very small compared to disks, and using a large amount of
+  memory will affect the normal operation of the node, so you must be very
+  careful to use the minimum amount necessary.
+
+If you use memory backed `emptyDir` without appropriate settings, e.g. without
```
Comment (thockin) [inline content/en/docs/concepts/configuration/manage-resources-containers.md:273]: minor: "If you use a memory backed `emptyDir` without appropriate limits, one or more of the following may occur:"

File: content/en/docs/concepts/configuration/manage-resources-containers.md [hunk 19]
```diff
@@ -237,6 +238,73 @@ are available in your cluster, then Pod resource usage can be retrieved either
...
+To prevent these risks, everyone who can define a Pod should specify memory
+limits for pod or `sizeLimit` for `emptyDir`.
+
+Since memory limits apply to all pod memory usage, and the `sizeLimit` only
+applies to the `emptyDir`, a memory limits is the proper tool to prevent a
+container from consuming too much of the node's allocatable memory.
+The `sizeLimit` is really a tool to prevent usage of the memory-backed `emptyDir`
+from causing the container to hit its memory limit and OOM, not a tool to limit
+a pod's memory uasge. So it is better concidering to specify memory limits for
+pod first.
```
Comment (thockin) [inline content/en/docs/concepts/configuration/manage-resources-containers.md:296]: I feel like all of this is expressed in the earlier text, and this is redundant - disagree?

File: content/en/docs/concepts/configuration/manage-resources-containers.md [hunk 20]
```diff
@@ -237,6 +238,73 @@ are available in your cluster, then Pod resource usage can be retrieved either
...
+
+If you are administering a cluster or namespace, you also set
```
Comment (thockin) [inline content/en/docs/concepts/configuration/manage-resources-containers.md:298]: ..you can also set...

File: content/en/docs/concepts/configuration/manage-resources-containers.md [hunk 21]
```diff
@@ -237,6 +238,73 @@ are available in your cluster, then Pod resource usage can be retrieved either
...
+[ResourceQuota](/docs/concepts/policy/resource-quotas/) that limits memory use;
+you may also want to define a [LimitRange](/docs/concepts/policy/limit-range/)
+for additional enforcement.
+If you specify a `spec.containers[].resources.limits.memory` for each Pod,
+then the default size of an `emptyDir` volume will be `.spec.resources.limits.memory`.
```
Comment (thockin) [inline content/en/docs/concepts/configuration/manage-resources-containers.md:303]: change "default size" to "maximum size" ?

Comment (thockin) [reply content/en/docs/concepts/configuration/manage-resources-containers.md:303]: Change "`.spec.resources.limits.memory`" to "the pod's memory limit"   (you already spelled out the field path before)

File: content/en/docs/concepts/storage/volumes.md [hunk 22]
```diff
@@ -255,11 +255,17 @@ If that is filled up from another source (for example, log files or image
 overlays), the `emptyDir` may run out of capacity before this limit.
 
 {{< note >}}
-If the `SizeMemoryBackedVolumes` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) is enabled,
-you can specify a size for memory backed volumes.  If no size is specified, memory
-backed volumes are sized to node allocatable memory.
+You can specify a size for memory backed volumes, provided that the `SizeMemoryBackedVolumes`
+[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
+enabled in your cluster (this has been beta, and active by default, since the Kubernetes 1.22 release).
```
Comment (thockin) [inline content/en/docs/concepts/storage/volumes.md:260]: ...is enabled

**Comment by shu-mutou** [comment]
@thockin Thank you for your review! :bow: 
I'll fix them in few days!!

File: content/en/docs/concepts/configuration/manage-resources-containers.md (hunk 13 above)
Comment (shu-mutou) [reply content/en/docs/concepts/configuration/manage-resources-containers.md:246]: Yes! But I'd like to note that having multiple emptyDir increases the risk of OOM significantly. So I'd like to add following:
```
It is possible to create any number of `emptyDir`s that could potentially consume all available memory on the node, making OOM more likely.
```

File: content/en/docs/concepts/configuration/manage-resources-containers.md (hunk 15 above)
Comment (shu-mutou) [reply content/en/docs/concepts/configuration/manage-resources-containers.md:262]: changed to "things like language-level garbage collection.", thanks!

File: content/en/docs/concepts/configuration/manage-resources-containers.md (hunk 19 above)
Comment (shu-mutou) [reply content/en/docs/concepts/configuration/manage-resources-containers.md:296]: I'd like to remove all of this and above one.

**Comment by shu-mutou** [comment]
@thockin Thank you for your valuable reviews. I have made some fixes, so I hope you will take a look at them again. :bow:

File: content/en/docs/concepts/configuration/manage-resources-containers.md (hunk 13 above)
Comment (thockin) [reply content/en/docs/concepts/configuration/manage-resources-containers.md]: SGTM

**Comment by thockin** [comment]
Thanks!

/label tide/merge-method-squash
/lgtm
/approve

**Comment by shu-mutou** [comment]
@sftim Could you approve this?

**Review by nate-double-u — COMMENTED** [review]
Thanks @shu-mutou, this is a great update.

/approve
